sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java

```diff
@@ -0,0 +1,314 @@
+      getRequestBuilder
+          .getStateKeyBuilder()
+          .getOrderedListUserStateBuilder()
+          .getRangeBuilder()
+          .setStart(minTimestamp.getMillis())
+          .setEnd(limitTimestamp.getMillis());
+
+      // TODO: consider use cache here
+      CachingStateIterable<TimestampedValue<T>> persistentValues =
+          StateFetchingIterators.readAllAndDecodeStartingFrom(
+              Caches.noop(),
+              this.beamFnStateClient,
+              getRequestBuilder.build(),
+              this.timestampedValueCoder);
+
+      // Make a snapshot of the current pendingRemoves and use them to filter persistent values.
+      // The values of pendingRemoves are kept, so that they will still be accessible in
+      // pre-existing iterables even after a sort key is removed.
+      TreeRangeSet<Instant> pendingRemovesSnapshot = TreeRangeSet.create(pendingRemoves);
+      Iterable<TimestampedValue<T>> persistentValuesAfterRemoval =
+          Iterables.filter(
+              persistentValues, v -> !pendingRemovesSnapshot.contains(v.getTimestamp()));
+
+      return Iterables.mergeSorted(
+          ImmutableList.of(persistentValuesAfterRemoval, valuesInRange),
+          Comparator.comparing(TimestampedValue::getTimestamp));
+    }
+
+    return valuesInRange;
+  }
+
+  public Iterable<TimestampedValue<T>> read() {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    return readRange(Instant.ofEpochMilli(Long.MIN_VALUE), Instant.ofEpochMilli(Long.MAX_VALUE));
+  }
+
+  public void clearRange(Instant minTimestamp, Instant limitTimestamp) {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    // Remove items (in a collection) in the specific range from pendingAdds.
+    // The old values of the removed sub map are kept, so that they will still be accessible in
+    // pre-existing iterables even after the sort key is cleared.
+    pendingAdds.subMap(minTimestamp, limitTimestamp).clear();
+    if (!isCleared) {
+      pendingRemoves.add(
+          Range.range(minTimestamp, BoundType.CLOSED, limitTimestamp, BoundType.OPEN));
+    }
+  }
+
+  public void clear() {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+    isCleared = true;
+    // Create a new object for pendingRemoves and clear the mappings in pendingAdds.
+    // The entire tree range set of pendingRemoves and the old values in the pendingAdds are kept,
+    // so that they will still be accessible in pre-existing iterables even after the state is
+    // cleared.
+    pendingRemoves = TreeRangeSet.create();
+    pendingAdds.clear();
+  }
+
+  public void asyncClose() throws Exception {
+    isClosed = true;
+
+    if (!pendingRemoves.isEmpty()) {
```

Comment:
Do we need some kind of check `if (isCleared)`, where we first send a state request deleting the range [int64min, int64max)? If so, make sure to add a test.

Reply:
Nice catch! That's a case I should have handled.

Reply:
Test added.